src/groupeddataframe/splitapplycombine.jl

```diff
@@ -962,14 +1056,51 @@ function _combine(f::AbstractVector{<:Pair},
                 throw(ArgumentError("a single value or vector result is required when passing " *
                                     "multiple functions (got $(typeof(res)))"))
             end
+            if !(firstres isa AbstractVector) && isnothing(idx_agg)
+                idx_agg = Vector{Int}(undef, length(gd))
+                fillfirst!(nothing, idx_agg, 1:length(gd.groups), gd)
+            end
             idx, outcols, _ = _combine_with_first(wrap(firstres), fun, gd, incols,
-                                                  Val(firstmulticol))
+                                                  Val(firstmulticol),
+                                                  firstres isa AbstractVector ? nothing : idx_agg)
+            @assert length(outcols) == 1
             res[i] = idx, outcols[1]
         end
     end
-    # TODO: avoid recomputing idx for each pair when !all(isagg, f)
-    idx = res[1][1]
+    idx_loc = findfirst(x -> x[1] !== idx_agg, res)
+    if isnothing(idx_loc)
+        @assert !isnothing(idx_agg)
+        idx = idx_agg
+    else
+        idx = res[idx_loc][1]
+        agg2idx_map = nothing
+        for i in 1:length(res)
+            if res[i][1] != idx
```

Comment:
Maybe this would we worth it? Surprisingly, `==` doesn't use that optimization.
```suggestion
            if res[i][1] !== idx && res[i][1] != idx
```

Reply:
Yes - I think that I even used this trick somewhere in the code, but forgotten to add it here. Probably a PR to Base would be in place.